agent/grpc-external/services/resource/delete.go

```diff
@@ -80,6 +89,44 @@ func (s *Server) Delete(ctx context.Context, req *pbresource.DeleteRequest) (*pb
 	}
 }
 
+// Create a tombstone to capture the intent to delete child resources.
+// Tombstones are created preemptively to prevent partial failures even though
+// we are currently unaware of the success/failure/no-op of DeleteCAS. In
+// the failure and no-op cases the tombstone is effectively a no-op and will
+// still be deleted from the system by the reaper controller.
+func (s *Server) maybeCreateTombstone(ctx context.Context, deleteId *pbresource.ID, deleteVersion string) error {
+	// Don't create a tombstone when the resource being deleted is itself a tombstone.
+	if proto.Equal(resource.TypeV1Tombstone, deleteId.Type) {
+		return nil
+	}
+
+	data, err := anypb.New(&pbresource.Tombstone{
+		OwnerId:      deleteId,
+		OwnerVersion: deleteVersion,
+	})
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed creating tombstone: %v", err)
+	}
+
+	tombstone := &pbresource.Resource{
+		Id: &pbresource.ID{
+			Type:    resource.TypeV1Tombstone,
+			Tenancy: deleteId.Tenancy,
+			// TODO(spatel): Does this need to be more unique (embed deleteId.Uid)?
+			Name: fmt.Sprintf("%s-tombstone", deleteId.Name),
+		},
+		Data: data,
+		Metadata: map[string]string{
+			"generated_at": time.Now().Format(time.RFC3339),
+		},
+	}
+
+	if _, err := s.Write(ctx, &pbresource.WriteRequest{Resource: tombstone}); err != nil {
+		return fmt.Errorf("failed writing tombstone: %w", err)
+	}
+	return nil
+}
```

Comment:
I'm not seeing where the `ErrCASFailure` occurs in the following scenario. 

Given an existing resource "foo":

Scenario:
1. User instructs ResourceService to Delete("foo")
2. ResourceService creates tombstone via WriteCAS("tombstone-foo-uid1")
3. ResourceService fails to DeleteCAS("foo") and returns error to User
4. Time passes and User retries Delete("foo")
6a. Branch a: tombstone has not been reconciled by the reaper controller and still exists
    - ResourceService creates tombstone via WriteCAS("tombstone-foo-uid1")
    - WriteCAS() is identical to the write in Step 2. and thus a no-op.
    - Delete("foo") successful and scenario ends. 
6b. Branch b: tombstone has been reconciled by the reaper controller and has been deleted
    - ResourceService creates a tombstone via WriteCAS("tombstone-foo-uid1")
    - WriteCAS() creates a new "tombstone-foo-uid1".
    - Delete("foo") is successful and the scenario ends.

Reply:
> WriteCAS() is identical to the write in Step 2. and thus a no-op.

This is where we'll get an `ErrCASFailure` ☝🏻 because we're providing a blank version.

Reply:
Just spotted where my confusion was coming from, see: https://github.com/hashicorp/consul/pull/17108#issuecomment-1525436560